node-authorizer/pkg/authorizers/gce/identity.go

```diff
@@ -0,0 +1,328 @@
+	Email           string                `json:"email"`
+	EmailVerified   bool                  `json:"email_verified"`
+	Google          gceIdentityDataGoogle `json:"google"`
+}
+
+type gceIdentityDataGoogle struct {
+	ComputeEngine gceIdentityDataComputeEngine `json:"compute_engine"`
+}
+
+type gceIdentityDataComputeEngine struct {
+	ProjectID                 string `json:"project_id"`
+	ProjectNumber             int64  `json:"project_number"`
+	Zone                      string `json:"zone"`
+	InstanceID                string `json:"instance_id"`
+	InstanceName              string `json:"instance_name"`
+	InstanceCreationTimestamp int64  `json:"instance_creation_timestamp"`
+}
+
+const AudienceNodeBootstrap = "node-bootstrap.kubernetes.io"
+
+// getLocalInstanceIdentityClaim reads and returns the signed GCE instance identity description document
+func getLocalInstanceIdentityClaim(ctx context.Context, audience string) (string, error) {
+	client := &http.Client{}
+
+	v := url.Values{}
+
+	v.Set("audience", audience)
+	format := "full"
+	v.Set("format", format)
+
+	u := "http://metadata/computeMetadata/v1/instance/service-accounts/default/identity?" + v.Encode()
+	req, err := http.NewRequest("GET", u, nil)
+	if err != nil {
+		return "", fmt.Errorf("error building GCE identity document request: %v", err)
+	}
+
+	req = req.WithContext(ctx)
+
+	req.Header.Add("Metadata-Flavor", "Google")
+	resp, err := client.Do(req)
+	if err != nil {
+		return "", fmt.Errorf("error retrieving GCE identity document from metadata: %v", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("unexpected status %d reading GCE identity document from metadata", resp.StatusCode)
+	}
+
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", fmt.Errorf("error reading GCE identity document from metadata: %v", err)
+	}
+	return string(b), nil
+}
+
+type jwtHeader struct {
+	Algorithm string `json:"alg"`
+	KeyID     string `json:"kid"`
+}
+
+type cachedKeys struct {
+	raw       []byte
+	timestamp time.Time
+
+	mutex sync.Mutex
+	certs map[string]*x509.Certificate
+}
+
+func (c *cachedKeys) buildKeys() error {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if c.certs != nil {
+		return nil
+	}
+
+	certMap := make(map[string]string)
+	if err := json.Unmarshal(c.raw, &certMap); err != nil {
+		return fmt.Errorf("error parsing certificates: %v", err)
+	}
+
+	certs := make(map[string]*x509.Certificate)
+	for k, v := range certMap {
+		block, _ := pem.Decode([]byte(v))
+		if block == nil {
+			return fmt.Errorf("failed to parse certificate PEM")
+		}
+
+		cert, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			return fmt.Errorf("failed to parse x509 certificate: %v", err)
+		}
+
+		certs[k] = cert
+	}
+
+	c.certs = certs
+	return nil
+}
+
+type Validator struct {
+	certURL string
+
+	mutex sync.Mutex
+	keys  *cachedKeys
+}
+
+func NewValidator(ctx context.Context) (*Validator, error) {
```

Comment:
You might be able to reuse https://github.com/coreos/go-oidc/blob/v2/verify.go#L46-L70 here .. which will handle the key rotation for you .. That's for openid but the validation "should" be the same ..

Reply:
Thanks - I'll look at how that does it, but it does seem like it pulls in lots of code for what should be fairly simple.  Of course, if we implement selective key rotation here, it may be a wash :-)